apiserver/facades/client/application/update_series.go

```diff
@@ -0,0 +1,185 @@
+	// given machine. This includes all the subordinates.
+	Application(string) (Application, error)
+}
+
+// UpdateSeriesValidator defines an application validator.
+type UpdateSeriesValidator interface {
+	// ValidateApplication attempts to validate an application for
+	// a given series. Using force to allow the overriding of the error to
+	// ensure all application validate.
+	//
+	// I do question if you actually need to validate anything if force is
+	// employed here?
+	ValidateApplication(application Application, series string, force bool) error
+}
+
+// UpdateSeriesAPI provides the update series API facade for any given version.
+// It is expected that any API parameter changes should be performed before
+// entering the API.
+type UpdateSeriesAPI struct {
+	state     UpdateSeriesState
+	validator UpdateSeriesValidator
+}
+
+// NewUpdateSeriesAPI creates a new UpdateSeriesAPI
+func NewUpdateSeriesAPI(
+	state UpdateSeriesState,
+	validator UpdateSeriesValidator,
+) *UpdateSeriesAPI {
+	return &UpdateSeriesAPI{
+		state:     state,
+		validator: validator,
+	}
+}
+
+func (a *UpdateSeriesAPI) UpdateSeries(tag, series string, force bool) error {
+	if series == "" {
+		return errors.BadRequestf("series missing from args")
+	}
+	applicationTag, err := names.ParseApplicationTag(tag)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	app, err := a.state.Application(applicationTag.Id())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !app.IsPrincipal() {
+		return &params.Error{
+			Message: fmt.Sprintf("%q is a subordinate application, update-series not supported", applicationTag.Id()),
+			Code:    params.CodeNotSupported,
+		}
+	}
+
+	if err := a.validator.ValidateApplication(app, series, force); err != nil {
+		return errors.Trace(err)
+	}
+
+	return app.UpdateApplicationSeries(series, force)
+}
+
+type updateSeriesValidator struct {
+	localValidator  UpdateSeriesValidator
+	removeValidator UpdateSeriesValidator
+}
+
+func makeUpdateSeriesValidator(client CharmhubClient) updateSeriesValidator {
+	return updateSeriesValidator{
+		localValidator: stateSeriesValidator{},
+		removeValidator: charmhubSeriesValidator{
+			client: client,
+		},
+	}
+}
+
+func (s updateSeriesValidator) ValidateApplication(app Application, series string, force bool) error {
+	// This is not a charmhub charm, so we can fallback to querying state
+	// for the supported series.
+	if origin := app.CharmOrigin(); origin == nil || !corecharm.CharmHub.Matches(origin.Source) {
+		return s.localValidator.ValidateApplication(app, series, force)
+	}
+
+	return s.removeValidator.ValidateApplication(app, series, force)
+}
+
+// stateSeriesValidator validates an application using the state (database)
+// version of the charm.
+type stateSeriesValidator struct{}
+
+// ValidateApplications attempts to validate a series of applications for
+// a given series.
+func (s stateSeriesValidator) ValidateApplication(application Application, series string, force bool) error {
```

Comment:
These were the validations common between the machine-manager and application APIs right?

If we have the right indirections for dependencies, are they suitable for putting in a core package?

Reply:
I did think about this, but I opted to not this time. Also, we could place the interfaces and strategy into core, but the actual implementation of local and remote would be better served elsewhere. I could add a comment to that fact.

The general problem here and why I didn't generalize this, was because we model based on the API (bulk) and not our own modeling. Essentially we don't require bulk operations in this facade, yet we do in the other and trying to shoehorn both into one, isn't ideal.

I was thinking about abstracting this up into a library and dropping it somewhere, which I might agree to, but I'm loathed to put into core.